FIAT/dual_set.py

```diff
@@ -54,13 +57,67 @@ def to_riesz(self, poly_set):
         tshape = self.nodes[0].target_shape
         num_nodes = len(self.nodes)
         es = poly_set.get_expansion_set()
+        ed = poly_set.get_embedded_degree()
         num_exp = es.get_num_members(poly_set.get_embedded_degree())
+        nbf = poly_set.get_num_members()
 
         riesz_shape = tuple([num_nodes] + list(tshape) + [num_exp])
 
         self.mat = numpy.zeros(riesz_shape, "d")
 
-        for i in range(len(self.nodes)):
-            self.mat[i][:] = self.nodes[i].to_riesz(poly_set)
+        # let's amalgamate the pt_dict and deriv_dicts of all the
+        # functionals so we can tabulate the basis functions twice only
+        # (once on pts and once on derivatives)
+
+        # Need: dictionary mapping pts to which functionals they come from
+        pts_to_ells = collections.OrderedDict()
+        dpts_to_ells = collections.OrderedDict()
+
+        for i, ell in enumerate(self.nodes):
+            for pt in ell.pt_dict:
+                if pt in pts_to_ells:
+                    pts_to_ells[pt].append(i)
+                else:
+                    pts_to_ells[pt] = [i]
+
+        for i, ell in enumerate(self.nodes):
+            for pt in ell.deriv_dict:
+                if pt in dpts_to_ells:
+                    dpts_to_ells[pt].append(i)
+                else:
+                    dpts_to_ells[pt] = [i]
+
+        # Now tabulate
+        pts = list(pts_to_ells.keys())
+        bfs = es.tabulate(ed, pts)
```

Comment:
`bfs`?

Reply:
basis functions.  Expanded to "expansion_values"